base/applications/kbswitch/kbswitch.c

```diff
@@ -35,52 +42,158 @@ UINT ShellHookMessage = 0;
+    DWORD dwSize, dwIndex;
+    HKEY hKey, hLayoutKey;
+
+    g_cSpecialIds = 0;
+
+    if (RegOpenKeyEx(HKEY_LOCAL_MACHINE,
+                     TEXT("SYSTEM\\CurrentControlSet\\Control\\Keyboard Layouts"),
+                     0, KEY_READ, &hKey) != ERROR_SUCCESS)
     {
-        dwBufLen = sizeof(szTempLCID);
-        dwRes = RegQueryValueEx(hKey, szLayoutNum, NULL, NULL, (LPBYTE)szTempLCID, &dwBufLen);
-        if (dwRes != ERROR_SUCCESS)
+        return;
+    }
+
+    for (dwIndex = 0; dwIndex < 1000; ++dwIndex)
+    {
+        dwSize = ARRAYSIZE(szKLID);
+        if (RegEnumKeyEx(hKey, dwIndex, szKLID, &dwSize, NULL, NULL, NULL, NULL) != ERROR_SUCCESS)
+            break;
+
+        if (RegOpenKeyEx(hKey, szKLID, 0, KEY_READ, &hLayoutKey) != ERROR_SUCCESS)
+            continue;
+
+        dwSize = sizeof(szLayoutId);
+        if (RegQueryValueEx(hLayoutKey, TEXT("Layout Id"), NULL, NULL,
+                            (LPBYTE)szLayoutId, &dwSize) == ERROR_SUCCESS)
         {
-            RegCloseKey(hKey);
-            return FALSE;
+            DWORD dwKLID = _tcstoul(szKLID, NULL, 16);
+            WORD wLangId = LOWORD(dwKLID), wLayoutId = LOWORD(_tcstoul(szLayoutId, NULL, 16));
+            HKL hKL = (HKL)(LONG_PTR)(SPECIAL_MASK | MAKELONG(wLangId, wLayoutId));
+
+            /* Add a special ID */
+            g_SpecialIds[g_cSpecialIds].dwLayoutId = wLayoutId;
+            g_SpecialIds[g_cSpecialIds].hKL = hKL;
+            StringCchCopy(g_SpecialIds[g_cSpecialIds].szKLID,
+                          ARRAYSIZE(g_SpecialIds[g_cSpecialIds].szKLID), szKLID);
+            ++g_cSpecialIds;
         }
 
-        RegCloseKey(hKey);
+        RegCloseKey(hLayoutKey);
+
+        if (g_cSpecialIds >= ARRAYSIZE(g_SpecialIds))
+        {
+            OutputDebugStringA("g_SpecialIds is full!");
+            break;
+        }
+    }
+
+    RegCloseKey(hKey);
+}
+
+static VOID
+GetKLIDFromHKL(HKL hKL, LPTSTR szKLID, SIZE_T KLIDLength)
+{
+    szKLID[0] = 0;
+
+    if (IS_IME_HKL(hKL))
+    {
+        StringCchPrintf(szKLID, KLIDLength, _T("%08lx"), (DWORD)(DWORD_PTR)hKL);
+        return;
     }
 
-    /* Look for a substitute of this layout */
-    if (RegOpenKeyEx(HKEY_CURRENT_USER, _T("Keyboard Layout\\Substitutes"), 0,
-                     KEY_QUERY_VALUE, &hKey) == ERROR_SUCCESS)
+    if (IS_SPECIAL_HKL(hKL))
     {
-        dwBufLen = sizeof(szTempLCID);
-        if (RegQueryValueEx(hKey, szTempLCID, NULL, NULL, (LPBYTE)szLCID, &dwBufLen) != ERROR_SUCCESS)
+        INT i;
+        for (i = 0; i < g_cSpecialIds; ++i)
         {
-            /* No substitute found, then use the old LCID */
-            StringCchCopy(szLCID, LCIDLength, szTempLCID);
+            if (g_SpecialIds[i].hKL == hKL)
+            {
+                StringCchCopy(szKLID, KLIDLength, g_SpecialIds[i].szKLID);
+                return;
+            }
         }
+    }
+    else
+    {
+        StringCchPrintf(szKLID, KLIDLength, _T("%08lx"), LOWORD(hKL));
```

Comment:
If for example French keyboard is installed under English language then you have to use `HIWORD(hKL)` to get proper French KLID from HKL:

`StringCchPrintf(szKLID, KLIDLength, _T("%08lx"), HIWORD(hKL) != 0 ? HIWORD(hKL) : LOWORD(hKL));`

See tests https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/InputLanguageTests.cs#L113-L120